suite/tests/client-interface/tls.dll.cpp

```diff
@@ -176,9 +176,51 @@ event_exit(void)
     drmgr_exit();
 }
 
+static void
+test_calloc(void)
+{
+#ifndef WINDOWS
+    /* FIXME: Investigate why this test does not pass on Windows. */
+    uint previously_allocated = 0;
+    for (int i = 0; i < 100; i++) {
+        uint total_allocated = 0;
+        uint min_offset = (uint)-1;
+        /* Randomly allocate until an unaligned allocation of
+         * a single slot fails, meaning space is exhausted.
+         */
+        for (;;) {
+            uint num_slots = rand() % 4 + 1;         /* 1, 2, 3, 4 */
+            uint alignment = 1 << (rand() % 4) >> 1; /* 0, 1, 2, 4 */
+            reg_id_t reg;
+            uint offset;
+            if (dr_raw_tls_calloc(&reg, &offset, num_slots, alignment)) {
+                ASSERT(alignment == 0 || ALIGNED(offset, alignment));
+                total_allocated += num_slots;
+                if (offset < min_offset)
+                    min_offset = offset;
+            } else {
+                if (num_slots == 1 && alignment <= 1)
+                    break;
+            }
+        }
+        ASSERT(total_allocated > 0);
+        if (previously_allocated == 0)
+            previously_allocated = total_allocated;
+        else
+            ASSERT(previously_allocated == total_allocated);
+        /* Free the slots. */
+        ASSERT(dr_raw_tls_cfree(min_offset, total_allocated));
```

Comment:
This freeing of multiple allocations in one call is not going to work on Windows where the slots from different calloc calls are *not* guaranteed to be adjacent to each other: there are multiple sources of slots and they are not contiguous.  This is an API usage error and probably explains the `ntdll.c:1529 taken[i]` assert I see in the prior GA CI run.

Please add an explicit note to the dr_raw_tls_calloc() docs in dr_tools.h to help clarify that it cannot be assumed that multiple calls will return slots from the same region, nor can it be assumed that this client itself is the only entity reserving such slots (so that no one else tries to use multiple allocation calls but a single free call).

Reply:
I am now freeing each allocation separately.